src/torrent.cpp

```diff
@@ -5291,12 +5291,20 @@ namespace libtorrent
 		TORRENT_ASSERT(std::count(m_peers_to_disconnect.begin()
 			, m_peers_to_disconnect.end(), p) == 0);
 
-		std::weak_ptr<torrent> weak_t = shared_from_this();
-		m_peers_to_disconnect.push_back(p);
-		m_deferred_disconnect.post(m_ses.get_io_service(), [=]() {
-			std::shared_ptr<torrent> t = weak_t.lock();
-			if (t) t->on_remove_peers();
-		});
+		// we only schedule the peer for actual removal if in fact
+		// it is attached to this torrent, since this method can
+		// be called from the attach_peer path and fail to do so
+		// because of too many connections
+		if (p->associated_torrent().lock().get() == this)
```

Comment:
at first, I did it like your idea in 1) but was like "too much magic" in `torrent::remove_peer`, but once the problem is clear, I will redo it that way.

about 2), I'm not sure, I like the idea of immediate destruction if not needed, if you keep a `shared_ptr` they will stored in `m_undead_peers` for next tick. Maybe for another PR.